hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -311,25 +311,58 @@ case class TableParallelize(rowsAndGlobal: IR, nPartitions: Option[Int] = None)
 
   protected[ir] override def execute(ctx: ExecuteContext): TableValue = {
     val hc = HailContext.get
-    val (ptype, res) = CompileAndEvaluate._apply(ctx, rowsAndGlobal, optimize = false) match {
-      case Right((t, off)) => (t.fields(0).typ, SafeRow(t, off).getAs[Row](0))
+    val (ptype: PStruct, res) = CompileAndEvaluate._apply(ctx, rowsAndGlobal, optimize = false) match {
+      case Right((t: PTuple, off)) => (t.fields(0).typ, t.loadField(off, 0))
     }
 
-    val Row(_rows: IndexedSeq[_], globals: Row) = res
-
-    val rows = _rows.asInstanceOf[IndexedSeq[Row]]
-    rows.zipWithIndex.foreach { case (r, idx) =>
-      if (r == null)
-        fatal(s"cannot parallelize null values: found null value at index $idx")
+    val globalsT = ptype.types(1).asInstanceOf[PStruct]
+    val globals = BroadcastRow(RegionValue(ctx.r, ptype.loadField(res, 1)), globalsT, ctx.backend)
+
+    val rowsT = ptype.types(0).asInstanceOf[PArray]
+    val rowT = rowsT.elementType.asInstanceOf[PStruct]
+    val spec = TypedCodecSpec(rowT, BufferSpec.wireSpec)
+
+    val makeEnc = spec.buildEncoder(rowT)
+    val rowsAddr = ptype.loadField(res, 0)
+    val nRows = rowsT.loadLength(rowsAddr)
+
+    val parts = partition(nRows, nPartitions.getOrElse(16))
```

Comment:
I still think you should min with nRows.

Reply:
oops, I had a brain fart and assumed `partition` was doing that. Fixed.